[Issue #25] Added ScanBasedSourceOperator

sandeepreddy602 opened the pull request:
@chenlica @JavierJia ... Implemented ScanBasedSourceOperator and refactored a few things. I changed the design a little by making the interfaces IScanBasedSourceOperator and IIndexBasedSourceOperator as classes. Please review the code and please let me know if anything needs to be changed.

JavierJia commented on textdb/pom.xml:
```diff
@@ -1,61 +1,63 @@
+            <dependency>
+                <groupId>org.apache.lucene</groupId>
+                <artifactId>lucene-core</artifactId>
+                <version>5.5.0</version>
```
want to replace it with `${lucene.version}`?

JavierJia commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/constants/LuceneConstants.java:
```diff
@@ -0,0 +1,14 @@
+ *
+ */
+public class LuceneConstants {
+    public static final String INDEX_DIR = "index";
```
why do we use these value as constants? they are more like the configuration values.

sandeepreddy602 replied:
The Lucene constants will be used in many places. So, I had put it in a different class. If we want to make them configuration values, it will be easy if they are defined at one place.

JavierJia commented:
looks very good to me. :+1:

chenlica commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/source/SampleDataStore.java:
```diff
@@ -0,0 +1,107 @@
+/**
+ * 
+ */
+package edu.uci.ics.textdb.dataflow.source;
```
How is this class related to "ScanBasedSourceOperator"?

sandeepreddy602 replied:
SampleDataStore is a sample class to store the data in the Lucene index. ScanBasedSourceOperator scans this data. So they are actually not related. Ideally we need to put the actual DataStore classes in textdb-storage project, which I assumed we will be doing at later point.

JavierJia replied:
so will you make it a better interface + implementation in a separate PR?

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/source/IndexSearchSourceOperator.java:
```diff
@@ -0,0 +1,28 @@
+package edu.uci.ics.textdb.dataflow.source;
```
Is it better to include this "IndexSearchSourceOperator.java" in the next pull request to keep the modularity?

sandeepreddy602 replied:
IIndexSeachSourceOperator interface was already there in the master. Changed it to a class now. So it became part of this pull request.

chenlica commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/source/SampleDataStoreTest.java:
```diff
@@ -0,0 +1,27 @@
+/**
```
Should we have a "ScanBasedSourceOperatorTest.java"?

sandeepreddy602 replied:
Currently RegexMatcherTest tests ScanBasedSourceOperator class also. But that is not unit testing. Will add Unit test cases for ScanBasedSourceOperator.

chenlica commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/regexmatch/RegexMatcherTest.java:
```diff
@@ -1,24 +1,59 @@
+    
+    @Before
+    public void setUp() throws Exception{
+        SampleDataStore dataStore = new SampleDataStore();
```
How is the "SampleDataStore" object related to the "ScanBasedSourceOperator" used later?

sandeepreddy602 replied:
Here I am populating the Lucene data base before scanning the data.

chenlica commented:
Let's discuss this pull request when we meet on Friday.  We will make the structure right, with the right class names and test cases.

sandeepreddy602 commented:
@chenlica @JavierJia ..1.  Added ScanBasedSourceOperatorTest class to test scan operator …
2. Updated textdb pom to use lucene.version property
3. Cleaning up the data in Lucene after the test case is run. 
Please review the changes.

chenlica commented:
I will let Jianfeng review first if possible.  I will also do a review on the weekend.

JavierJia commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/source/ScanBasedSourceOperatorTest.java:
```diff
@@ -0,0 +1,58 @@
+    
+    @After
+    public void cleanUp() throws IOException{
+        dataStore.clearData();
```
do we need to `close` it?

sandeepreddy602 replied:
I am clearing the data after running the test, since it is a independent test case and should leave no trails.

JavierJia commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/common/DataTuple.java:
```diff
@@ -1,48 +1,55 @@
+/**
+ * 
+ */
+package edu.uci.ics.textdb.dataflow.common;
```
want to move this tuple to `main`?

sandeepreddy602 replied:
@JavierJia .. you mean different package or different project ( to textdb-common)?

JavierJia replied:
I made a mistake. You've already done it.

JavierJia commented:
LGTM :+1:  I left some minor comments. The only concern is that we should extract the required interface from the `SampleDataStore` and put it into the `textdb-storage` project. However, it could be finished in a separate PR.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/source/ScanBasedSourceOperator.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public void close() throws DataFlowException {
+        cursor = -1;
```
Per our discussion, it will be good to define a constant for "-1" within this class.

sandeepreddy602 replied:
will make these changes.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/source/ScanBasedSourceOperator.java:
```diff
@@ -0,0 +1,113 @@
+            Document document = indexSearcher.doc(scoreDocs[cursor++].doc);
+            List<IField> fields = new ArrayList<IField>();
+            for (String fieldName : schema) {
+                //TODO Generalize it to all field types(int, date etc)
```
Can you do another PR to do this generalization to support other types, including INT, DOUBLE, and DATE?

sandeepreddy602 replied:
sure.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/source/ScanBasedSourceOperator.java:
```diff
@@ -0,0 +1,113 @@
+            indexReader = DirectoryReader.open(directory);
+            indexSearcher = new IndexSearcher(indexReader);
+            //TODO we don't need analyzer for scanning all the docs!!
+            Analyzer analyzer = new StandardAnalyzer();
```
Check if we can avoid this analyzer.  If so, figure it out a way.  If not, let's remove this TODO.

sandeepreddy602 replied:
Looks like there is no other way of constructing the QueryParser object without passing an Analyzer. Will remove the TODO

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/source/ScanBasedSourceOperator.java:
```diff
@@ -0,0 +1,113 @@
+            indexSearcher = new IndexSearcher(indexReader);
+            //TODO we don't need analyzer for scanning all the docs!!
+            Analyzer analyzer = new StandardAnalyzer();
+            //TODO ideally default field should be combination of all the fields
```
What should we do for this TODO?

sandeepreddy602 replied:
Default field is not applicable to ScanBasedSourceOperator. It is applicable to IndexBasedSourceOperator. Will remove the TODO

chenlica commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/regexmatch/RegexMatcherTest.java:
```diff
@@ -1,24 +1,67 @@
+        ITuple nextTuple = null;
+        int counter = 0;
+        while((nextTuple = regexMatcher.getNextTuple()) != null){
+            assertEquality(TestConstants.SAMPLE_TUPLES.get(counter++), nextTuple);
```
This assert assumes that the order of the returned tuples is the same as the order in which they were inserted.  This assumption may not be true.  I suggest we check: (1) if a new tuple is contained in the set of known tuples; and (2) the total number of returned tuples is the same as the known tuples.

sandeepreddy602 replied:
Will make these changes

chenlica commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/source/SampleDataStore.java:
```diff
@@ -0,0 +1,107 @@
+ * @author sandeepreddy602
+ *
+ */
+public class SampleDataStore {
```
Instead of using hard-coded constants (tuples and schema), can we pass them to SampleDataStore as parameters?  In this way, we can easily have different kinds of sample data stores by passing different parameters.  In the ScanBasedSourceOperator test cases, we can easily do more tests with different sample data sources.

sandeepreddy602 replied:
Will move SampleDataStore to textdb-store project and will make these changes in a different PR

chenlica commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/source/SampleDataStoreTest.java:
```diff
@@ -0,0 +1,27 @@
+    }
+    
+    @Test
+    public void testStoreData() throws Exception{
```
This test is really doing nothing.  Can you improve it?

sandeepreddy602 replied:
Will create a separate PR for Data Store related classes

chenlica commented:
I added a few more comments.  Please take care of them.

sandeepreddy602 commented:
@chenlica @JavierJia .. Please review the changes. Will make the changes related to data store and mutiple fields support in the next PRs.

JavierJia commented:
LGTM

chenlica commented on textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/source/ScanBasedSourceOperatorTest.java:
```diff
@@ -0,0 +1,72 @@
+        ITuple nextTuple = null;
+        int numTuples = 0;
+        while((nextTuple  = scanBasedSourceOperator.getNextTuple()) != null){
+            //Checking if the tuple retieved is present in the samplesTuples
```
Minor typo "retieved" :-)

chenlica commented:
I made a minor comment.  Please fix it and do the merge.  

It will be good to do the other PRs quickly since we want other teams to use this operator to come up with their first scan-based implementation this week.  

Thank you!
